File: packages/server/src/domains/users/service.ts
import bcrypt from 'bcrypt';
import type { User, CreateUserInput, UpdateUserInput, UpdateProfileInput, SessionUser } from '@flick/shared';
import * as repo from './repo.js';
import { NotFoundError, ConflictError, UnauthorizedError, ForbiddenError, ValidationError } from '../../shared/errors.js';
import { authConfig } from '../../shared/config.js';

const BCRYPT_ROUNDS = 12;

function toSessionUser(user: User): SessionUser {
  return { id: user.id, email: user.email, name: user.name, role: user.role, auth_method: user.auth_method };
}

export async function listUsers(): Promise<User[]> {
  return repo.findAll();
}

export async function getUser(id: string): Promise<User> {
  const user = await repo.findById(id);
  if (!user) throw new NotFoundError('User', id);
  return user;
}

export async function getProfile(id: string): Promise<User> {
  const user = await repo.findById(id);
  if (!user) throw new NotFoundError('User', id);
  return user;
}

export async function createUser(input: CreateUserInput): Promise<User> {
  if (authConfig.mode === 'saml') {
    throw new ForbiddenError('Cannot create password users when auth mode is SAML-only');
  }

  const existing = await repo.findByEmail(input.email);
  if (existing) throw new ConflictError(`User with email '${input.email}' already exists`);

  const password_hash = await bcrypt.hash(input.password, BCRYPT_ROUNDS);
  return repo.create({ ...input, password_hash });
}

export async function updateUser(id: string, input: UpdateUserInput): Promise<User> {
  const user = await repo.update(id, input);
  if (!user) throw new NotFoundError('User', id);
  return user;
}

export async function updateProfile(id: string, input: UpdateProfileInput): Promise<SessionUser> {
  if (input.email) {
    const existing = await repo.findByEmail(input.email);
    if (existing && existing.id !== id) {
      throw new ConflictError(`User with email '${input.email}' already exists`);
    }
  }

  const user = await repo.updateProfile(id, input);
  if (!user) throw new NotFoundError('User', id);
  return toSessionUser(user);
}

export async function changePassword(id: string, currentPassword: string, newPassword: string): Promise<void> {
  const user = await repo.findByIdWithHash(id);
  if (!user) throw new NotFoundError('User', id);

  if (user.auth_method !== 'password') {
    throw new ForbiddenError('SSO users cannot change password');
  }

  const valid = await bcrypt.compare(currentPassword, user.password_hash!);
  if (!valid) throw new UnauthorizedError('Current password is incorrect');

  const hash = await bcrypt.hash(newPassword, BCRYPT_ROUNDS);
  await repo.updatePasswordHash(id, hash);
}

export async function adminResetPassword(adminId: string, targetId: string, newPassword: string): Promise<void> {
  const target = await repo.findByIdWithHash(targetId);
  if (!target) throw new NotFoundError('User', targetId);

  if (target.auth_method !== 'password') {
    throw new ForbiddenError('Cannot reset password for SSO users');
  }

  const hash = await bcrypt.hash(newPassword, BCRYPT_ROUNDS);
  await repo.updatePasswordHash(targetId, hash);
}

export async function deleteUser(id: string): Promise<void> {
  const deleted = await repo.remove(id);
  if (!deleted) throw new NotFoundError('User', id);
}

export async function login(email: string, password: string): Promise<SessionUser> {
  if (authConfig.mode === 'saml') {
    throw new ForbiddenError('Password login is disabled. Use SSO to sign in.');
  }

  const user = await repo.findByEmail(email);
  if (!user) throw new UnauthorizedError('Invalid email or password');

  if (user.auth_method === 'saml') {
    throw new ForbiddenError('This account uses SSO. Please sign in with your identity provider.');
  }

  const valid = await bcrypt.compare(password, user.password_hash!);
  if (!valid) throw new UnauthorizedError('Invalid email or password');

  return toSessionUser(user);
}

export interface SamlProfile {
  nameId: string;
  issuer: string;
  email: string;
  name: string;
}

export async function loginWithSaml(profile: SamlProfile): Promise<SessionUser> {
  // Try to find by SAML identity first
  const existing = await repo.findBySamlIdentity(profile.nameId, profile.issuer);
  if (existing) {
    // Update name/email if changed in IdP
    const updated = await repo.updateProfile(existing.id, { name: profile.name, email: profile.email });
    return toSessionUser(updated ?? existing);
  }

  // Check for email collision with password user
  const emailUser = await repo.findByEmail(profile.email);
  if (emailUser) {
    throw new ConflictError(
      `An account with email '${profile.email}' already exists with password authentication. Contact an administrator.`,
    );
  }

  // JIT provisioning: create new SAML user
  const newUser = await repo.createSamlUser({
    email: profile.email,
    name: profile.name,
    saml_name_id: profile.nameId,
    saml_issuer: profile.issuer,
  });

  return toSessionUser(newUser);
}

File: packages/server-rust/src/domains/users/service.rs
use serde::Serialize;
use sqlx::PgPool;
use uuid::Uuid;

use crate::config::{AppConfig, AuthMode};
use crate::error::AppError;

use super::repo;

// ── Types ──────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct SessionUser {
    pub id: Uuid,
    pub email: String,
    pub name: String,
    pub role: String,
    pub auth_method: String,
}

impl From<&repo::UserWithHashRow> for SessionUser {
    fn from(u: &repo::UserWithHashRow) -> Self {
        Self {
            id: u.id,
            email: u.email.clone(),
            name: u.name.clone(),
            role: u.role.clone(),
            auth_method: u.auth_method.clone(),
        }
    }
}

impl From<&repo::UserRow> for SessionUser {
    fn from(u: &repo::UserRow) -> Self {
        Self {
            id: u.id,
            email: u.email.clone(),
            name: u.name.clone(),
            role: u.role.clone(),
            auth_method: u.auth_method.clone(),
        }
    }
}

pub struct CreateUserInput {
    pub email: String,
    pub name: String,
    pub password: String,
    pub role: String,
}

pub struct UpdateUserInput {
    pub name: Option<String>,
    pub email: Option<String>,
    pub role: Option<String>,
}

pub struct UpdateProfileInput {
    pub name: Option<String>,
    pub email: Option<String>,
}

pub struct SamlProfile {
    pub name_id: String,
    pub issuer: String,
    pub email: String,
    pub name: String,
}

// ── Service functions ──────────────────────────────────────────────────────

pub async fn list_users(db: &PgPool) -> Result<Vec<repo::UserRow>, AppError> {
    repo::find_all(db).await.map_err(AppError::from)
}

pub async fn get_user(db: &PgPool, id: Uuid) -> Result<repo::UserRow, AppError> {
    repo::find_by_id(db, id)
        .await?
        .ok_or_else(|| AppError::not_found("User", Some(&id.to_string())))
}

pub async fn get_profile(db: &PgPool, id: Uuid) -> Result<repo::UserRow, AppError> {
    repo::find_by_id(db, id)
        .await?
        .ok_or_else(|| AppError::not_found("User", Some(&id.to_string())))
}

pub async fn create_user(
    db: &PgPool,
    config: &AppConfig,
    input: CreateUserInput,
) -> Result<repo::UserRow, AppError> {
    if config.auth_mode == AuthMode::Saml {
        return Err(AppError::validation(
            "Cannot create password users when auth mode is SAML",
        ));
    }

    // Check email conflict
    if repo::find_by_email(db, &input.email).await?.is_some() {
        return Err(AppError::conflict("A user with this email already exists"));
    }

    let password_hash =
        bcrypt::hash(&input.password, 12).map_err(|e| AppError::Internal(anyhow::anyhow!(e.to_string())))?;

    repo::create(db, &input.email, &input.name, &password_hash, &input.role)
        .await
        .map_err(AppError::from)
}

pub async fn update_user(
    db: &PgPool,
    id: Uuid,
    input: UpdateUserInput,
) -> Result<repo::UserRow, AppError> {
    repo::update(
        db,
        id,
        input.name.as_deref(),
        input.email.as_deref(),
        input.role.as_deref(),
    )
    .await?
    .ok_or_else(|| AppError::not_found("User", Some(&id.to_string())))
}

pub async fn update_profile(
    db: &PgPool,
    id: Uuid,
    input: UpdateProfileInput,
) -> Result<repo::UserRow, AppError> {
    // Check email conflict with a different user
    if let Some(ref email) = input.email {
        if let Some(existing) = repo::find_by_email(db, email).await? {
            if existing.id != id {
                return Err(AppError::conflict("A user with this email already exists"));
            }
        }
    }

    repo::update_profile(db, id, input.name.as_deref(), input.email.as_deref())
        .await?
        .ok_or_else(|| AppError::not_found("User", Some(&id.to_string())))
}

pub async fn change_password(
    db: &PgPool,
    id: Uuid,
    current_password: &str,
    new_password: &str,
) -> Result<(), AppError> {
    let user = repo::find_by_id_with_hash(db, id)
        .await?
        .ok_or_else(|| AppError::not_found("User", Some(&id.to_string())))?;

    if user.auth_method != "password" {
        return Err(AppError::validation(
            "Password change is only available for password-based accounts",
        ));
    }

    let hash = user
        .password_hash
        .ok_or_else(|| AppError::Internal(anyhow::anyhow!("No password hash found")))?;

    let valid =
        bcrypt::verify(current_password, &hash).map_err(|e| AppError::Internal(anyhow::anyhow!(e.to_string())))?;

    if !valid {
        return Err(AppError::unauthorized("Current password is incorrect"));
    }

    let new_hash =
        bcrypt::hash(new_password, 12).map_err(|e| AppError::Internal(anyhow::anyhow!(e.to_string())))?;

    repo::update_password_hash(db, id, &new_hash).await?;

    Ok(())
}

pub async fn admin_reset_password(
    db: &PgPool,
    target_id: Uuid,
    new_password: &str,
) -> Result<(), AppError> {
    let user = repo::find_by_id_with_hash(db, target_id)
        .await?
        .ok_or_else(|| AppError::not_found("User", Some(&target_id.to_string())))?;

    if user.auth_method != "password" {
        return Err(AppError::validation(
            "Password reset is only available for password-based accounts",
        ));
    }

    let new_hash =
        bcrypt::hash(new_password, 12).map_err(|e| AppError::Internal(anyhow::anyhow!(e.to_string())))?;

    repo::update_password_hash(db, target_id, &new_hash).await?;

    Ok(())
}

pub async fn delete_user(db: &PgPool, id: Uuid) -> Result<(), AppError> {
    let deleted = repo::remove(db, id).await?;
    if !deleted {
        return Err(AppError::not_found("User", Some(&id.to_string())));
    }
    Ok(())
}

pub async fn login(
    db: &PgPool,
    config: &AppConfig,
    email: &str,
    password: &str,
) -> Result<SessionUser, AppError> {
    if config.auth_mode == AuthMode::Saml {
        return Err(AppError::validation(
            "Password login is disabled when auth mode is SAML",
        ));
    }

    let user = repo::find_by_email(db, email)
        .await?
        .ok_or_else(|| AppError::unauthorized("Invalid email or password"))?;

    if user.auth_method != "password" {
        return Err(AppError::unauthorized("Invalid email or password"));
    }

    let hash = user
        .password_hash
        .as_deref()
        .ok_or_else(|| AppError::unauthorized("Invalid email or password"))?;

    let valid = bcrypt::verify(password, hash).map_err(|e| AppError::Internal(anyhow::anyhow!(e.to_string())))?;

    if !valid {
        return Err(AppError::unauthorized("Invalid email or password"));
    }

    Ok(SessionUser::from(&user))
}

pub async fn login_with_saml(db: &PgPool, profile: SamlProfile) -> Result<SessionUser, AppError> {
    // Try to find by SAML identity first
    if let Some(user) =
        repo::find_by_saml_identity(db, &profile.name_id, &profile.issuer).await?
    {
        return Ok(SessionUser::from(&user));
    }

    // Check if a user with this email already exists (collision)
    if let Some(existing) = repo::find_by_email(db, &profile.email).await? {
        // Email collision with a non-SAML user
        if existing.auth_method != "saml" {
            return Err(AppError::conflict(
                "A user with this email already exists with a different auth method",
            ));
        }
        // SAML user exists with same email but different name_id/issuer — treat as conflict
        return Err(AppError::conflict(
            "A SAML user with this email already exists with a different identity",
        ));
    }

    // JIT provision: create new SAML user
    let user =
        repo::create_saml_user(db, &profile.email, &profile.name, &profile.name_id, &profile.issuer)
            .await?;

    Ok(SessionUser::from(&user))
}
